README
Games-Tournament-Swiss

Version 0.15

The complexity of the C11 code slowed this release down. C12 code wasn't touched, so I'm sure bugs remain. 

This is a major revision. In addition, we have a nicer interface to writing pairing tests. Look at three.t in the t/ directory.

Warning!

This is ALPHA-level software. Although it attempts to implement the FIDE Swiss Rules, it may not pair players correctly. It may take over your computer, and force you to reboot. It may DELETE files. It may DESTROY your computer. Use at your own risk.

The complexity (recursiveness) of the C10-14 procedures and the bugs hidden there means nothing can be guaranteed. Cross your fingers that your tournament does not exercise the C12 code!

NOTICE: More tests are needed. If you have an interesting pairing table that is difficult to pair, please report it as a bug at http://search.cpan.org/~drbean/Games-Tournament-Swiss/, ideally in the same format as t/three.t. Bring it on! Try to break Games::Tournament::Swiss and report your results!

I recommend looking at the scripts. If you want some more help, look at the 
tests. Again, scripts are still likely to break, but the purpose of that is to get you to write your own scripts. (ha-ha) Look at the scripts and the tests.

This program is run from a shell. You need to work from a command line and type commands. It also depends on perl to run. If you don't know anything about perl, you will have to learn something about installing it and running scripts, before you can use this software.

Alternatively, there is a web app frontend in a subversion repo at http://svn.openfoundry.org/swiss

For perl on Windows, try the free download of ActiveState's distribution at http://www.activestate.com/store/activeperl, or the experimental Strawberry Perl http://win32.perl.org/wiki/index.php?title=Strawberry_Perl

On other platforms, find it from your regular source for new packages.

INSTALLATION

This software doesn't install itself. To install this module, run the following commands:

    perl Build.PL		or		perl Makefile.PL
    ./Build					make
    ./Build test				make install
    ./Build install

make may be nmake or dmake on Windows. The tests will take a few minutes.

The distribution is made up of modules and scripts. The modules provide functionality that you can use to write your own tournament software. The scripts are an example of a swiss pairing program.

SCRIPTS

The scripts in script_files/ need to be installed somewhere so that they can be run in the directory in which pairing of each round is done.

DIRECTORY LAYOUT

The scripts assume that there is a directory in which a configuration file, called league.yaml, with data about the players exists. The rounds are paired in subdirectories, named 1,2,3,.. below this directory. Files named:

brackets.yaml  pairing.yaml  round.yaml
matches.yaml   player.yaml   tourney.yaml

may be left in these subdirectories after pairing.

Scores are entered in the same subdirectories in a file called 'scores.yaml'. You need to create these files, if you use 'pairstately'.

You can create a pairing table with 'pairingtable,' convert it into YAML form with 'pairtable2yaml' and run 'pair.'

DATA FILES

Do not use tabs in these YAML files. The level of indentation is significant. Follow the examples closely. The first, league.yaml has lines of the form:

member:
  - id: 1
    name: Laver, Rod
    rating: 2810
    title: Grandmaster
  - id: 2
    name: Sampras, Pete
    rating: 2800
    title: Unknown
  - id: 3
    name: McEnroe, John
    rating: 2780
    title: Unknown

If you are using your own scoring scheme, and colors (called, roles), see the example below. You can add your own data to the member records. A pairing number is generated for the players, so don't add a pairing number.

Late entries in a round after the first get their own "late" records, of the same form as the member records.

TODO Configuration of your own scoring scheme looks like it is broken.

The score files: 1.yaml, etc, are of the form:

'Laver, Rod': Loss
'Sampras, Pete': Win
'McEnroe, John': Bye

First number is the name of the player. (Every name will have to be unique.) The second is the result for that player.

Alternatively,
1:
  - 'Laver, Rod': 0
  - 'Sampras, Pete': 1
2: 
  - 'McEnroe, John': -

A template file is generated in the rounds subdirectory.

A sample scoring scheme, abbreviations and roles in league.yaml could be:

scores:
  win: 1
  loss: 0.5
  draw: 0.75
  absent: 0
  bye: 1
abbreviation:
  W: White
  1: Win
  '=': Draw
roles:
  - White
  - Black

For the form of the pairing table and its YAML equivalent, see the documentation for 'pairtable2yaml.'

GENERATING PAIRINGS

Starting with an empty main directory, create league.yaml, and an empty subdirectory for the first round. Run the script, 'pair' in the empty round subdirectory. A log of the pairing is printed and 'round.yaml' in the directory contains the matches. After the games in the round are complete, in the scores subdirectory, enter the scores for the players in the file, '1.yaml', or whatever the round is. You can use the file called 1.yaml, or whatever in the round subdirectory as a template and afterward move it to the scores subdirectory. Then you can run 'crosstable 1' or 'pairingtable 1' in the original directory above the subdirectory, to get current standings. Convert your pairing table into YAML form with 'pairtable2yaml' and copy it into a subdirectory for the next round, and continue as before.

Alternatively, set up the Catalyst web app frontend and access Games::Tournament::Swiss through the Internet.

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc Games::Tournament::Swiss

You can also look for information at:

    Search CPAN, for documentation and source:
        http://search.cpan.org/dist/Games-Tournament-Swiss

    CPAN Request Tracker (Report Bugs here):
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Tournament-Swiss

    AnnoCPAN, annotated CPAN documentation:
        http://annocpan.org/dist/Games-Tournament-Swiss

    CPAN Ratings:
        http://cpanratings.perl.org/d/Games-Tournament-Swiss

COPYRIGHT AND LICENCE

Copyright (C) 2010 Dr Bean

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
---
version: 0.18
date: 六 12月  4 13:08:01 CST 2010
revision: about 1095 at openfoundry, 6440 on my computer
changes:
  - rt 29495 flipped back
  - script score.yaml files in same round directory as round.yaml
  - cleaned up handling of Byes with a strict following of FIDE Rules in particular in C13 in connection with rt 62502
  - Contestant::Swiss floats method
  - Card hasPlayer method

---
version: 0.17
date: 2009å¹´  7æ 24æ¥ éææ¥ 12:28:38 JST
revision: about 110
changes:
  - Late entries. Test: lateentry.t
  - rt 48419 b is not q-w. http://chesschat.org/showpost.php?p=248965&postcount
=42. Test: x.t

---
version: 0.16
date: 二  6月 30 13:40:45 CST 2009
revision: about 1700
changes:
  - rt 47450 E5 A7c S2 bug: players were being assigned preferences
  - t/initializePreferences.t regression test

---
version: 0.15
date: Nov 28, 2007
revision: about 1645
changes:
  - Preference update tweaking, tests
  - Config provides defaults overrideable in BEGIN block
  - breadth-first C11 searching with X inheritance by remainder groups
  - more attractive pairing test writer interface

---
version: 0.12
date: Oct 27, 2007
revision: about 1536
changes:
  - greater understanding of proactive float checking in C6
  - large C11, C12 changes
  - prepareCards method makes Cards from hashes
  - loggedProcedures, log, tailLog for mix-&-match messaging

---
version: 0.11
date: Oct 13, 2007
revision: about 1480
changes:
  - rewrite of C10 code
  - new indexing method
  - table ordering in round.yaml bartolin #29953

---
version: 0.10
date: Oct 11, 2007
revision: about 1470
changes:
  - hashification of bracket, match orders, with a thisBracket index
  - BYE procedure gets own state
  - bad floating over-engineering checked
  - transposition restart after C12,13 refloating
  - test of c6pairs: bartolin #29682

---
version: 0.09
date: Sep 30, 2007
revision: about 1431
changes:
  - refactor of C6 float testing
  - D1,2 transposition, exchange tests
  - pairing table consistency checks
  - pairingtable parsing: bartolin #29485
  - unit test of C10 code: bartolin #29495

---
version: 0.08
date: Sep 16, 2007
revision: about 1374
changes:
  - reorder S2 before C8 swapping: bartolin #29258
  - C1 tests of B1, B2 incompatibility were not cooperating.
  - separate remainder group pairing recording: #29073
  - round number passing/guessing in cross/pairingtable scripts: bartolin #29125
  - pairstately supports abbreviation, scores grouped by table: bartolin #28834
  - test that pair, pairstately, pairingtable, pairtable2yaml work

---
version: 0.07
date: Sep 4, 2007
revision: about 1334
changes:
  - refactored FIDE.pm to test pairings in C6PAIRS for all p possibilities, rather than just first pprime ones.
  - reworked C13, C14 to handle http://www.chesschat.org/showthread.php?t=6619, with cc6619 test.

---
version: 0.06
date: Sep 1, 2007
revision: about 1316
changes:
  - pairtable2yaml parses pairing table, emits YAML
  - pair uses YAML pairing table instead of score files, YAML serialization files.
  - Fixed bug #29023, test with t/foura.t 

---
version: 0.05
date: August 2007
revision: about 1300
changes:
  - Make crosstable, pairingtable also use letter scores: #28834
  - Copying evacuees by reference meant when they exited group they were gone from the copy as well: #28797.
  - Scripts take game results from file with lines of form 'Capablanca, Jose Raul': Win
  - provide traditional Makefile.PL also

---
version: 0.04
date: April 2007
changes:
  - Pair script wants letter scores, not numbers.
  - Test FIDE.pm pairs correctly in 4-entrant tournament.

---
version: 0.03
date: February 2007
changes:
  - First version released on CPAN.

---
version: 0.02
date: January 2007
changes:
  - First version, worked up after talking with Bill Gletsos on www.chesschat.org


lib/Games/Tournament/Card.pm
package Games::Tournament::Card;
BEGIN {
  $Games::Tournament::Card::VERSION = '0.18';
}

# Last Edit: 2010 12月 04, 15時36分05秒
# $Id: $

use warnings;
use strict;
use Carp;

use List::Util qw/min reduce sum first/;
use List::MoreUtils qw/any all/;
use Scalar::Util qw/looks_like_number/;

use constant ROLES => @Games::Tournament::Swiss::Config::roles?
			@Games::Tournament::Swiss::Config::roles:
			Games::Tournament::Swiss::Config->roles;

=head1 NAME

Games::Tournament::Card - A record of the results of a match

=cut

=head1 SYNOPSIS

    $game = Games::Tournament:Card->new(round => 1, contestants => {Black => $knicks, White => $deepblue}, result => { Black => 'Win', White => 'Loss' });

=head1 DESCRIPTION

In a tournament, matches take place in rounds between contestants, who are maybe floated, and who have roles, and there is a result for these matches, which can be written on a card. 

=head1 METHODS

=head2 new

    $game = Games::Tournament:Card->new(
	    round => 1,
	    contestants => {Black => $knicks, White => $deepblue},
	    result => { Black => 'Win', White => 'Loss' },
	    floats => { Black => 'Up', White => 'Down' }, or 
	    floats => { Black => 'Not', White => 'Not' }
    );
    $bye = Games::Tournament:Card->new(
	    round => 1,
	    contestants => {Bye => $player},
	    result => "Bye"
	    floats => 'Down' );

'contestants' is a hash ref of player objects, keyed on Black and White, or Home and Away, or some other role distinction that needs to be balanced over the tournament. The players are probably instances of the Games::Tournament::Contestant::Swiss class. 'result' is a hash reference, keyed on the same keys as contestants, containing the results of the match. 'floats' is a hash of  which role was floated up and which down. The default is neither contestant was floated, and 'Down' for a Bye. A4. What are the fields in Forfeits and byes? Forfeit and Tardy have no special form, other than { White => 'Forfeit', Black => 'Tardy' }. Bye is { Bye => $player }. TODO Perhaps the fields should be Winner and Loser, and Down and Up?

=cut 

sub new {
    my $self = shift;
    my %args = @_;
    return bless \%args, $self;
}


=head2 canonize

    $game->canonize

Fleshes out a partial statement of the result. From an abbreviated match result (eg, { Black => 'Win' }), works out a canonical representation (eg, { Black => 'Win', White => 'Loss' }). A bye result is represented as { Bye => 'Bye' }.

=cut 

sub canonize {
    my $self        = shift;
    my $round       = $self->round;
    my $contestants = $self->contestants;
    my $result      = $self->result;
    my %result;
    my %roles = map { $contestants->{$_}->{id} => $_ } keys %$contestants;
    warn
"Incomplete match of @{[values( %roles )]} players @{[map {$_->id} values %$contestants]} in round $round.\n"
      unless keys %roles == 2
      or grep m/bye/i, values %roles;
  ROLE: foreach my $contestant ( values %$contestants ) {
        my $role = $roles{ $contestant->{id} };
        if ( $role eq 'Bye' ) {
                $result{$role} = $result->{$role} = 'Bye';
            }
        elsif ( exists $result->{$role} ) {
            if ( $result->{$role} =~ m/^(?:Win|Loss|Draw|Forfeit)$/i ) {
                $result{$role} = $result->{$role};
            }
            else {
                warn
"$result->{$role} result for player $contestant->{id} in round $round";
            }
            next ROLE;
        }
        elsif ( values %$contestants != 1 ) {
            my @opponents =
              grep { $contestant->id ne $_->id } values %$contestants;
            my $opponent = $opponents[0];
            my $other    = $roles{ $opponent->id };
            if ( exists $result->{$other} ) {
                $result{$role} = 'Loss'
                  if $result->{$other} =~ m/^Win$/i;
                $result{$role} = 'Win'
                  if $result->{$other} =~ m/^Loss$/i;
                $result{$role} = 'Draw'
                  if $result->{$other} =~ m/^Draw$/i;
            }
            else {
                warn
"$result->{$role}, $result->{$other} result for player $contestant->{id} and opponent $opponent->{id} in round $round";
            }
        }
	else {
		die "Not a Bye, but no result or no partner";
	}
    }
    $self->result( \%result );
}


=head2 myResult

    $game->myResult($player)

Returns the result for $player from $game, eg 'Win', 'Loss' or 'Draw'.
TODO Should return 0,0.5,1 in numerical context.

=cut 

sub myResult {
    my $self       = shift;
    my $contestant = shift;
    $self->canonize;
    my $contestants = $self->contestants;
    my $result      = $self->result;
    my %result;
    my %roles = map { $contestants->{$_}->id => $_ } keys %$contestants;
    my $role = $roles{ $contestant->id };
    return $result->{$role};
}


=head2 myPlayers

    $game->myPlayers

Returns an array of the players from $game, eg ($alekhine, $yourNewNicks) in ROLES order.

=cut 

sub myPlayers {
    my $self        = shift;
    my $contestants = $self->contestants;
    my @players;
    for my $role ( ROLES ) {
	push @players, $contestants->{$role} if exists $contestants->{$role};
    }
    push @players, $contestants->{Bye} if exists $contestants->{Bye};
    return @players;
}


=head2 hasPlayer

    $game->hasPlayer($player)

A predicate to perform a test to see if a player is a contestant in $game. Because different objects may refer to the same player when copied by value, use id to decide.

=cut 

sub hasPlayer {
    my $self        = shift;
    my $player = shift;
    my @contestants = $self->myPlayers;
    any { $player->id eq $_->id } @contestants;
}


=head2 myOpponent

    $game->myOpponent($player)

Returns the opponent of $player from $game. If $player has a Bye, return a Games::Tournament::Contestant::Swiss object with name 'Bye', and id 'Bye'.

=cut 

sub myOpponent {
    my $self       = shift;
    my $contestant = shift;
    my $id = $contestant->id;
    my $contestants = $self->contestants;
    my @contestants = values %$contestants;
    my %dupes;
    for my $contestant ( @contestants )
    {
	die "Player $contestant isn't a contestant"
	unless $contestant and
		$contestant->isa('Games::Tournament::Contestant::Swiss');
    }
    my @dupes = grep { $dupes{$_->id}++ } @contestants;
    croak "Players @dupes had more than one role" if @dupes;
    my $opponent = first { $id ne $_->id } @contestants;
    $opponent = Games::Tournament::Contestant::Swiss->new(
	name => "Bye", id => "Bye") if $self->isBye;
    return $opponent;
}


=head2 myRole

    $game->myRole($player)

Returns the role for $player from $game, eg 'White', 'Banker' or 'Away'.

=cut 

sub myRole {
    my $self       = shift;
    my $contestant = shift;
    my $id = $contestant->id;
    my $contestants = $self->contestants;
    my @contestants = $self->myPlayers;
    unless ( $self->hasPlayer($contestant) ) {
	my $round = $self->round;
	my $players;
	$players .= " $_: " . $contestants->{$_}->id for keys %$contestants;
	carp "Player $id not in Round $round match. Contestants are $players.";
	return;
    }
    my %dupes;
    for my $contestant ( @contestants )
    {
	die "Player $contestant isn't a contestant"
	unless $contestant and
		$contestant->isa('Games::Tournament::Contestant::Swiss');
    }
    my @dupes = grep { $dupes{$_->id}++ } @contestants;
    croak "Players @dupes had more than one role" if @dupes;
    my %roleReversal;
    for my $role ( keys %$contestants )
    {
	my $id = $contestants->{$role}->id;
	$roleReversal{$id} = $role;
    }
    my $role        = $roleReversal{ $id };
    carp "No role for player $id in round " . $self->round unless $role;
    return $role;
}


=head2 myFloat

    $game->myFloat($player)

Returns the float for $player in $game, eg 'Up', 'Down' or 'Not'.

=cut 

sub myFloat {
    my $self       = shift;
    my $contestant = shift;
    # $self->canonize;
    my $float = $self->float($contestant);
    return $float;
}


=head2 opponentRole

    Games::Tournament::Card->opponentRole( $role )

Returns the role of the opponent of the player in the given role. Class method.

=cut 

sub opponentRole {
    my $self       = shift;
    my $role = shift;
    my %otherRole;
    @otherRole{ (ROLES) } = reverse (ROLES);
    return $otherRole{ $role };
}


=head2 round

 $game->round

Returns the round in which the match is taking place.

=cut

sub round {
    my $self = shift;
    return $self->{round};
}


=head2 contestants

	$game->contestants

Gets/sets the participants as an anonymous array of player objects.

=cut

sub contestants {
    my $self        = shift;
    my $contestants = shift;
    if ( defined $contestants ) { $self->{contestants} = $contestants; }
    else { return $self->{contestants}; }
}


=head2 result

	$game->result

Gets/sets the results of the match.

=cut

sub result {
    my $self   = shift;
    my $result = shift;
    if ( defined $result ) { $self->{result} = $result; }
    else { return $self->{result}; }
}


=head2 equalScores

	$game->equalScores

Tests whether the players have equal scores, returning 1 or ''. If scores were not equal, they are (should be) floating.

=cut

sub equalScores {
    my $self   = shift;
    my $contestants = $self->contestants;
    my @score = map { $contestants->{$_}->score } ROLES;
    return unless looks_like_number $score[0];
    return all { $score[0] == $_ } @score;
}


=head2 higherScoreRole

	$game->higherScoreRole

Returns the role of the player with the higher score, returning '', if scores are equal.

=cut

sub higherScoreRole {
    my $self   = shift;
    my $contestant = $self->contestants;
    my @score = map { $contestant->{$_}->score } ROLES;
    return (ROLES)[0] if $score[0] > $score[1];
    return (ROLES)[1] if $score[0] < $score[1];
    return '';
}


=head2 floats

	$game->floats

Gets/sets the floats of the match. Probably $game->float($player, 'Up') is used however, instead.

=cut

sub floats {
    my $self   = shift;
    my $floats = shift;
    if ( defined $floats ) { $self->{floats} = $floats; }
    else { return $self->{floats}; }
}


=head2 float

	$card->float($player[,'Up|Down|Not'])

Gets/sets whether the player was floated 'Up', 'Down', or 'Not' floated. $player->floats is not changed. This takes place in $tourney->collectCards. TODO what if $player is 'Bye'?

=cut

sub float {
    my $self   = shift;
    my $player = shift;
    die "Player is $player ref"
      unless $player and $player->isa('Games::Tournament::Contestant::Swiss');
    my $role = $self->myRole($player);
    croak "Player " . $player->id . " has $role role in round $self->{round}?"
      unless $role eq 'Bye'
      or $role     eq (ROLES)[0]
      or $role     eq (ROLES)[1];
    my $float = shift;
    if ( $role eq 'Bye' ) { return 'Down'; }
    elsif ( defined $float ) { $self->{floats}->{$role} = $float; }
    elsif ( $self->{floats}->{$role} ) { return $self->{floats}->{$role}; }
    else { return 'Not'; }
}

=head2 isBye

	$card->isBye

Returns whether the card is for a bye rather than a game between two oppponents.

=cut

sub isBye {
    my $self   = shift;
    my $contestants = $self->contestants;
    my @status = keys %$contestants;
    return 1 if @status == 1 and any { $_ eq 'Bye' } @status;
    return 0 if @status == 2 and all { $_ eq (ROLES)[0] or $_ eq (ROLES)[1] } @status;
    return;
}

=head1 AUTHOR

Dr Bean, C<< <drbean, followed by the at mark (@), cpan, then a dot, and finally, org> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-games-tournament-match at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Tournament-Card>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Tournament::Card

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Tournament-Card>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Tournament-Card>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Tournament-Card>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Tournament-Card>

=back

=head1 ACKNOWLEDGEMENTS

=head1 COPYRIGHT & LICENSE

Copyright 2006 Dr Bean, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;    # End of Games::Tournament::Card

# vim: set ts=8 sts=4 sw=4 noet:
